File: reference/mathexpr/src/tokenizer.ts
/**
 * Converts a math expression string into a sequence of tokens.
 *
 * Supports: integers, decimals, operators (+, -, *, /, %, **), parentheses.
 * Whitespace is skipped. Throws on unrecognized characters.
 *
 * @node tokenizer
 * @depends-on token-types
 * @contract tokenizer.test.ts
 * @hint lexer: Walk the input character-by-character. No regex needed.
 */

import { type Token, token } from "./token-types";

export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;

  while (i < input.length) {
    const ch = input[i];

    if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
      i++;
      continue;
    }

    if (ch === "(") {
      tokens.push(token("lparen", "("));
      i++;
      continue;
    }

    if (ch === ")") {
      tokens.push(token("rparen", ")"));
      i++;
      continue;
    }

    if (ch === "+") {
      tokens.push(token("plus", "+"));
      i++;
      continue;
    }

    if (ch === "-") {
      tokens.push(token("minus", "-"));
      i++;
      continue;
    }

    if (ch === "*") {
      if (i + 1 < input.length && input[i + 1] === "*") {
        tokens.push(token("power", "**"));
        i += 2;
      } else {
        tokens.push(token("star", "*"));
        i++;
      }
      continue;
    }

    if (ch === "/") {
      tokens.push(token("slash", "/"));
      i++;
      continue;
    }

    if (ch === "%") {
      tokens.push(token("percent", "%"));
      i++;
      continue;
    }

    if (isDigit(ch) || ch === ".") {
      let num = "";
      let hasDot = false;
      while (i < input.length && (isDigit(input[i]) || input[i] === ".")) {
        if (input[i] === ".") {
          if (hasDot) {
            throw new Error(`Unexpected character '.' at position ${i}`);
          }
          hasDot = true;
        }
        num += input[i];
        i++;
      }
      tokens.push(token("number", num));
      continue;
    }

    throw new Error(`Unexpected character '${ch}' at position ${i}`);
  }

  return tokens;
}

function isDigit(ch: string): boolean {
  return ch >= "0" && ch <= "9";
}

File: research/experiments/sweep/runs/mathexpr-full-cpp-sonnet45/src/tokenizer.cpp
#include "tokenizer.h"
#include <stdexcept>
#include <cctype>

std::vector<Token> tokenize(const std::string& input) {
    std::vector<Token> tokens;
    size_t i = 0;

    while (i < input.size()) {
        char c = input[i];

        // Skip whitespace
        if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
            i++;
            continue;
        }

        // Numbers (digits or starting with '.')
        if (std::isdigit(c) || c == '.') {
            std::string num;
            bool hasDecimal = false;

            while (i < input.size()) {
                char ch = input[i];
                if (std::isdigit(ch)) {
                    num += ch;
                    i++;
                } else if (ch == '.') {
                    if (hasDecimal) {
                        throw std::runtime_error(std::string("Unexpected character '") + ch + "'");
                    }
                    hasDecimal = true;
                    num += ch;
                    i++;
                } else {
                    break;
                }
            }
            tokens.push_back(token(TokenKind::Number, num));
            continue;
        }

        // Check for ** (power)
        if (c == '*' && i + 1 < input.size() && input[i + 1] == '*') {
            tokens.push_back(token(TokenKind::Power, "**"));
            i += 2;
            continue;
        }

        // Single-character operators
        switch (c) {
            case '+':
                tokens.push_back(token(TokenKind::Plus, "+"));
                i++;
                break;
            case '-':
                tokens.push_back(token(TokenKind::Minus, "-"));
                i++;
                break;
            case '*':
                tokens.push_back(token(TokenKind::Star, "*"));
                i++;
                break;
            case '/':
                tokens.push_back(token(TokenKind::Slash, "/"));
                i++;
                break;
            case '%':
                tokens.push_back(token(TokenKind::Percent, "%"));
                i++;
                break;
            case '(':
                tokens.push_back(token(TokenKind::LParen, "("));
                i++;
                break;
            case ')':
                tokens.push_back(token(TokenKind::RParen, ")"));
                i++;
                break;
            default:
                throw std::runtime_error(std::string("Unexpected character '") + c + "' at position " + std::to_string(i));
        }
    }

    return tokens;
}
